File: backend/src/data/models/purchase-history.model.ts
import mongoose, { Document, Schema } from 'mongoose';
import { PurchaseHistory } from '../../domain/entities/purchase_history.entity';

export interface PurchaseHistoryDocument extends Document, Omit<PurchaseHistory, 'id' | 'offerId'> {
  _id: mongoose.Types.ObjectId;
  offerId: mongoose.Types.ObjectId;
}

const purchaseHistorySchema = new Schema<PurchaseHistoryDocument>({
  userId: { type: String, required: true },
  offerId: { type: Schema.Types.ObjectId, ref: 'Offer', required: true },
  purchaseDate: { type: Date, default: Date.now },
});

export const PurchaseHistoryModel = mongoose.model<PurchaseHistoryDocument>('PurchaseHistory', purchaseHistorySchema);

File: mobile_app/lib/data/models/purchase_history_model.dart
import '../../domain/entities/purchase_history.dart';
import 'offer_model.dart';

class PurchaseHistoryModel extends PurchaseHistory {
  PurchaseHistoryModel({
    required String id,
    required String userId,
    required OfferModel offer,
    required DateTime purchaseDate,
  }) : super(
          id: id,
          userId: userId,
          offer: offer,
          purchaseDate: purchaseDate,
        );

  factory PurchaseHistoryModel.fromJson(Map<String, dynamic> json) {
    return PurchaseHistoryModel(
      id: json['id'],
      userId: json['userId'],
      offer: OfferModel.fromJson(json['offer']),
      purchaseDate: DateTime.parse(json['purchaseDate']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'offer': (offer as OfferModel).toJson(),
      'purchaseDate': purchaseDate.toIso8601String(),
    };
  }
}